File: Assets/Scripts/Combat/TurnManager.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using KyndeBlade.Combat;

namespace KyndeBlade
{
    public enum CombatState
    {
        WaitingForInput,
        ExecutingAction,
        RealTimeWindow,
        ProcessingResults,
        CombatEnded
    }

    public class TurnManager : MonoBehaviour
    {
        [Header("Combat")]
        public float RealTimeWindowDuration = 2f;
        public GameSettings Settings;

        public CombatState State { get; private set; } = CombatState.CombatEnded;
        public List<MedievalCharacter> PlayerCharacters { get; private set; } = new List<MedievalCharacter>();
        public List<MedievalCharacter> EnemyCharacters { get; private set; } = new List<MedievalCharacter>();
        public List<MedievalCharacter> TurnOrder { get; private set; } = new List<MedievalCharacter>();
        public MedievalCharacter CurrentCharacter { get; private set; }
        public int TurnNumber { get; private set; }
        public float RealTimeWindowRemaining { get; private set; }
        public float ActionCastTimeRemaining { get; private set; }
        public float ActionExecutionTimeRemaining { get; private set; }
        /// <summary>Attacker whose strike is imminent (for strike-warning sound theme).</summary>
        public MedievalCharacter CurrentAttackerDuringWindow { get; private set; }

        CombatAction _currentAction;
        MedievalCharacter _currentTarget;

        public event Action<MedievalCharacter> OnTurnChanged;
        public event Action OnCombatEnded;
        /// <summary>Fired when an action is executed (executor, target, action). For AI observation.</summary>
        public event Action<MedievalCharacter, MedievalCharacter, CombatAction> OnActionExecuted;

        public void InitializeCombat(List<MedievalCharacter> players, List<MedievalCharacter> enemies)
        {
            PlayerCharacters = new List<MedievalCharacter>(players);
            EnemyCharacters = new List<MedievalCharacter>(enemies);
            TurnOrder = new List<MedievalCharacter>();
            TurnOrder.AddRange(PlayerCharacters);
            TurnOrder.AddRange(EnemyCharacters);
            TurnOrder.RemoveAll(c => c == null);
            CalculateTurnOrder();
        }

        public void AddCharacterToCombat(MedievalCharacter character, bool isPlayer)
        {
            if (character == null) return;
            if (isPlayer) PlayerCharacters.Add(character);
            else EnemyCharacters.Add(character);
            TurnOrder.Add(character);
            CalculateTurnOrder();
        }

        public void StartCombat()
        {
            if (TurnOrder.Count == 0) return;
            State = CombatState.WaitingForInput;
            TurnNumber = 1;
            CurrentCharacter = TurnOrder[0];
            CurrentCharacter?.OnTurnStart?.Invoke();
            OnTurnChanged?.Invoke(CurrentCharacter);
        }

        public void EndTurn()
        {
            CurrentCharacter?.OnTurnEnd?.Invoke();
        }

        public void NextTurn()
        {
            EndTurn();
            CheckCombatEnd();
            if (State == CombatState.CombatEnded) return;

            int idx = TurnOrder.IndexOf(CurrentCharacter);
            int start = (idx + 1) % TurnOrder.Count;
            int i = start;

            do
            {
                var next = TurnOrder[i];
                if (next != null && next.IsAlive())
                {
                    CurrentCharacter = next;
                    TurnNumber++;
                    State = CombatState.WaitingForInput;
                    CurrentCharacter?.OnTurnStart?.Invoke();
                    OnTurnChanged?.Invoke(CurrentCharacter);
                    return;
                }
                i = (i + 1) % TurnOrder.Count;
            } while (i != start);

            State = CombatState.CombatEnded;
            OnCombatEnded?.Invoke();
        }

        public void ExecuteAction(CombatAction action, MedievalCharacter target)
        {
            if (action == null || CurrentCharacter == null || State != CombatState.WaitingForInput) return;

            State = CombatState.ExecutingAction;
            _currentAction = action;
            _currentTarget = target;

            if (action.GetCastTime() > 0f)
                ActionCastTimeRemaining = action.GetCastTime();
            else
                ProcessActionCast();
        }

        void ProcessActionCast()
        {
            if (_currentAction == null || CurrentCharacter == null) return;

            CurrentCharacter.ExecuteCombatAction(_currentAction, _currentTarget);
            OnActionExecuted?.Invoke(CurrentCharacter, _currentTarget, _currentAction);

            if (_currentAction.GetExecutionTime() > 0f)
                ActionExecutionTimeRemaining = _currentAction.GetExecutionTime();
            else
                ProcessActionExecution();
        }

        void ProcessActionExecution()
        {
            if (_currentAction != null &&
                (_currentAction.ActionData.ActionType == CombatActionType.Escapade ||
                 _currentAction.ActionData.ActionType == CombatActionType.Ward))
            {
                var attacker = _currentTarget != null && EnemyCharacters.Contains(_currentTarget)
                    ? _currentTarget
                    : GetFirstAliveEnemy();
                StartRealTimeWindow(_currentAction.ActionData.SuccessWindow, attacker);
            }
            else
            {
                State = CombatState.ProcessingResults;
                Invoke(nameof(NextTurn), 0.01f);
            }

            _currentAction = null;
            _currentTarget = null;
            ActionCastTimeRemaining = 0f;
            ActionExecutionTimeRemaining = 0f;
        }

        public void StartRealTimeWindow(float duration, MedievalCharacter attacker = null)
        {
            if (Settings != null)
                duration = Settings.GetAdjustedWindow(duration);
            RealTimeWindowDuration = duration;
            RealTimeWindowRemaining = duration;
            CurrentAttackerDuringWindow = attacker ?? GetFirstAliveEnemy();
            State = CombatState.RealTimeWindow;
        }

        MedievalCharacter GetFirstAliveEnemy()
        {
            foreach (var e in EnemyCharacters)
                if (e != null && e.IsAlive()) return e;
            return null;
        }

        public bool IsPlayerTurn() => CurrentCharacter != null && PlayerCharacters.Contains(CurrentCharacter);

        void CalculateTurnOrder()
        {
            TurnOrder.Sort((a, b) =>
            {
                float sa = a != null ? a.Stats.Speed : 0f;
                float sb = b != null ? b.Stats.Speed : 0f;
                return sb.CompareTo(sa);
            });
        }

        void CheckCombatEnd()
        {
            if (AreAllDefeated(PlayerCharacters) || AreAllDefeated(EnemyCharacters))
            {
                State = CombatState.CombatEnded;
                OnCombatEnded?.Invoke();
            }
        }

        bool AreAllDefeated(List<MedievalCharacter> list)
        {
            foreach (var c in list)
                if (c != null && c.IsAlive()) return false;
            return true;
        }

        void Update()
        {
            if (State == CombatState.RealTimeWindow)
            {
                RealTimeWindowRemaining -= Time.deltaTime;
                if (RealTimeWindowRemaining <= 0f)
                {
                    State = CombatState.ProcessingResults;
                    NextTurn();
                }
                return;
            }

            if (State == CombatState.ExecutingAction)
            {
                if (ActionCastTimeRemaining > 0f)
                {
                    ActionCastTimeRemaining -= Time.deltaTime;
                    if (ActionCastTimeRemaining <= 0f) ProcessActionCast();
                }
                else if (ActionExecutionTimeRemaining > 0f)
                {
                    ActionExecutionTimeRemaining -= Time.deltaTime;
                    if (ActionExecutionTimeRemaining <= 0f) ProcessActionExecution();
                }
            }
        }
    }
}

File: ProjectArchive/legacy/unreal-5.7/Source/KyndeBlade/Combat/TurnManager.cpp
#include "TurnManager.h"
#include "../Characters/MedievalCharacter.h"
#include "CombatAction.h"
#include "Algo/Sort.h"

ATurnManager::ATurnManager()
{
	PrimaryActorTick.bCanEverTick = true;
	CombatState = ECombatState::CombatEnded;
	CurrentCharacter = nullptr;
	TurnNumber = 0;
	RealTimeWindowRemaining = 0.0f;
}

void ATurnManager::BeginPlay()
{
	Super::BeginPlay();
}

void ATurnManager::Tick(float DeltaTime)
{
	Super::Tick(DeltaTime);

	// Update real-time window
	if (CombatState == ECombatState::RealTimeWindow)
	{
		RealTimeWindowRemaining -= DeltaTime;
		if (RealTimeWindowRemaining <= 0.0f)
		{
			// Timeout: apply full pending damage if any (defender failed to dodge/parry)
			if (PendingDamageTarget && PendingDamageAttacker)
			{
				PendingDamageTarget->ApplyCustomDamage(PendingDamageAmount, PendingDamageAttacker);
				if (PendingBreakDamage > 0.0f)
				{
					PendingDamageTarget->TakeBreakDamage(PendingBreakDamage);
				}
			}
			PendingDamageAmount = 0.0f;
			PendingBreakDamage = 0.0f;
			PendingDamageTarget = nullptr;
			PendingDamageAttacker = nullptr;
			OnRealTimeWindowEnded.Broadcast();
			CombatState = ECombatState::ProcessingResults;
			NextTurn();
		}
		return;
	}

	// Process action timing (cast and execution phases)
	// Use else-if chain to ensure only one phase processes per frame
	if (CombatState == ECombatState::ExecutingAction)
	{
		// Process cast time first
		if (ActionCastTimeRemaining > 0.0f)
		{
			ActionCastTimeRemaining -= DeltaTime;
			if (ActionCastTimeRemaining <= 0.0f)
			{
				// Cast complete, move to execution phase
				ProcessActionCast();
				// Don't process execution in the same frame - it will be processed next frame
			}
		}
		// Only process execution if cast is complete (not in the same frame)
		else if (ActionExecutionTimeRemaining > 0.0f)
		{
			ActionExecutionTimeRemaining -= DeltaTime;
			if (ActionExecutionTimeRemaining <= 0.0f)
			{
				// Execution complete
				ProcessActionExecution();
			}
		}
	}
}

void ATurnManager::InitializeCombat(const TArray<AMedievalCharacter*>& InPlayerCharacters, const TArray<AMedievalCharacter*>& InEnemyCharacters)
{
	PlayerCharacters = InPlayerCharacters;
	EnemyCharacters = InEnemyCharacters;
	
	TurnOrder.Empty();
	TurnOrder.Append(PlayerCharacters);
	TurnOrder.Append(EnemyCharacters);

	// Remove null characters
	TurnOrder.RemoveAll([](AMedievalCharacter* Char) { return Char == nullptr; });

	CalculateTurnOrder();
}

void ATurnManager::AddCharacterToCombat(AMedievalCharacter* Character, bool bIsPlayer)
{
	if (!Character)
	{
		return;
	}

	if (bIsPlayer)
	{
		PlayerCharacters.AddUnique(Character);
	}
	else
	{
		EnemyCharacters.AddUnique(Character);
	}

	TurnOrder.AddUnique(Character);
	CalculateTurnOrder();
}

void ATurnManager::StartCombat()
{
	if (TurnOrder.Num() == 0)
	{
		return;
	}

	// Clear any existing timers
	GetWorld()->GetTimerManager().ClearTimer(NextTurnTimerHandle);

	CombatState = ECombatState::WaitingForInput;
	TurnNumber = 1;
	CurrentCharacter = TurnOrder[0];
	
	if (CurrentCharacter)
	{
		CurrentCharacter->OnTurnStart();
		OnTurnChanged.Broadcast(CurrentCharacter);
	}
}

void ATurnManager::EndTurn()
{
	if (CurrentCharacter)
	{
		CurrentCharacter->OnTurnEnd();
	}
}

void ATurnManager::NextTurn()
{
	EndTurn();

	// Check if combat should end
	CheckCombatEnd();
	if (CombatState == ECombatState::CombatEnded)
	{
		return;
	}

	// Find next alive character
	int32 CurrentIndex = TurnOrder.Find(CurrentCharacter);
	int32 NextIndex = (CurrentIndex + 1) % TurnOrder.Num();
	int32 StartIndex = NextIndex;

	do
	{
		AMedievalCharacter* NextChar = TurnOrder[NextIndex];
		if (NextChar && NextChar->IsAlive())
		{
			CurrentCharacter = NextChar;
			TurnNumber++;
			CombatState = ECombatState::WaitingForInput;
			
			if (CurrentCharacter)
			{
				CurrentCharacter->OnTurnStart();
				OnTurnChanged.Broadcast(CurrentCharacter);
			}
			return;
		}

		NextIndex = (NextIndex + 1) % TurnOrder.Num();
	} while (NextIndex != StartIndex);

	// No alive characters found - end combat
	GetWorld()->GetTimerManager().ClearTimer(NextTurnTimerHandle);
	CombatState = ECombatState::CombatEnded;
	OnCombatEnded.Broadcast();
}

void ATurnManager::ExecuteAction(UCombatAction* Action, AMedievalCharacter* Target)
{
	if (!Action || !CurrentCharacter || CombatState != ECombatState::WaitingForInput)
	{
		return;
	}

	CombatState = ECombatState::ExecutingAction;
	CurrentExecutingAction = Action;
	CurrentActionTarget = Target;

	// Handle cast time
	if (Action->GetCastTime() > 0.0f)
	{
		// Enter cast phase
		ActionCastTimeRemaining = Action->GetCastTime();
		ActionExecutionTimeRemaining = 0.0f;
		// Process cast time in Tick
	}
	else
	{
		// No cast time, execute immediately
		ProcessActionCast();
	}
}

void ATurnManager::ProcessActionCast()
{
	if (!CurrentExecutingAction || !CurrentCharacter)
	{
		return;
	}

	// For Strike vs player target, defer damage so defender gets real-time window
	const bool bDeferDamage = (CurrentExecutingAction->ActionData.ActionType == ECombatActionType::Strike
		&& CurrentActionTarget && PlayerCharacters.Contains(CurrentActionTarget));

	if (bDeferDamage)
	{
		CurrentCharacter->ExecuteCombatActionWithDeferredDamage(CurrentExecutingAction, CurrentActionTarget);
	}
	else
	{
		CurrentCharacter->ExecuteCombatAction(CurrentExecutingAction, CurrentActionTarget);
	}

	// Handle execution time
	if (CurrentExecutingAction->GetExecutionTime() > 0.0f)
	{
		ActionExecutionTimeRemaining = CurrentExecutingAction->GetExecutionTime();
		// Process execution time in Tick
	}
	else
	{
		// No execution time, finish immediately
		ProcessActionExecution();
	}
}

void ATurnManager::ProcessActionExecution()
{
	if (!CurrentExecutingAction)
	{
		return;
	}

	const ECombatActionType ActionType = CurrentExecutingAction->ActionData.ActionType;
	const bool bStrikeVsPlayer = (ActionType == ECombatActionType::Strike && CurrentActionTarget && PlayerCharacters.Contains(CurrentActionTarget));

	if (bStrikeVsPlayer)
	{
		// Store pending damage and give defender a real-time window
		float FinalDamage = CurrentExecutingAction->ActionData.Damage;
		if (CurrentActionTarget->IsBroken())
		{
			FinalDamage *= 1.5f;
		}
		PendingDamageAmount = FinalDamage;
		PendingBreakDamage = CurrentExecutingAction->ActionData.BreakDamage;
		PendingDamageTarget = CurrentActionTarget;
		PendingDamageAttacker = CurrentCharacter;
		StartRealTimeWindow(CurrentExecutingAction->ActionData.SuccessWindow > 0.0f ? CurrentExecutingAction->ActionData.SuccessWindow : 1.5f);
		OnRealTimeWindowStarted.Broadcast(CurrentActionTarget);
	}
	else if (ActionType == ECombatActionType::Escapade || ActionType == ECombatActionType::Ward)
	{
		StartRealTimeWindow(CurrentExecutingAction->ActionData.SuccessWindow);
		OnRealTimeWindowStarted.Broadcast(CurrentCharacter);
	}
	else
	{
		CombatState = ECombatState::ProcessingResults;
		GetWorld()->GetTimerManager().ClearTimer(NextTurnTimerHandle);
		GetWorld()->GetTimerManager().SetTimer(NextTurnTimerHandle, this, &ATurnManager::NextTurn, 0.01f, false);
	}

	CurrentExecutingAction = nullptr;
	CurrentActionTarget = nullptr;
	ActionCastTimeRemaining = 0.0f;
	ActionExecutionTimeRemaining = 0.0f;
}

void ATurnManager::StartRealTimeWindow(float Duration)
{
	RealTimeWindowDuration = Duration;
	RealTimeWindowRemaining = Duration;
	CombatState = ECombatState::RealTimeWindow;
}

void ATurnManager::NotifyDodgeSuccess(AMedievalCharacter* Defender)
{
	if (CombatState != ECombatState::RealTimeWindow || Defender != PendingDamageTarget)
	{
		return;
	}
	PendingDamageAmount = 0.0f;
	PendingBreakDamage = 0.0f;
	PendingDamageTarget = nullptr;
	PendingDamageAttacker = nullptr;
	OnRealTimeWindowEnded.Broadcast();
	CombatState = ECombatState::ProcessingResults;
	NextTurn();
}

void ATurnManager::NotifyParrySuccess(AMedievalCharacter* Defender)
{
	if (CombatState != ECombatState::RealTimeWindow || Defender != PendingDamageTarget)
	{
		return;
	}
	if (PendingDamageTarget && PendingDamageAttacker)
	{
		const float ReducedDamage = PendingDamageAmount * 0.3f;
		PendingDamageTarget->ApplyCustomDamage(ReducedDamage, PendingDamageAttacker);
		if (PendingBreakDamage > 0.0f)
		{
			PendingDamageTarget->TakeBreakDamage(PendingBreakDamage * 0.3f);
		}
	}
	PendingDamageAmount = 0.0f;
	PendingBreakDamage = 0.0f;
	PendingDamageTarget = nullptr;
	PendingDamageAttacker = nullptr;
	OnRealTimeWindowEnded.Broadcast();
	CombatState = ECombatState::ProcessingResults;
	NextTurn();
}

bool ATurnManager::IsPlayerTurn() const
{
	if (!CurrentCharacter)
	{
		return false;
	}

	return PlayerCharacters.Contains(CurrentCharacter);
}

void ATurnManager::CalculateTurnOrder()
{
	// Sort by speed (higher speed goes first)
	Algo::Sort(TurnOrder, [](const AMedievalCharacter* A, const AMedievalCharacter* B)
	{
		if (!A || !B) return false;
		return A->Stats.Speed > B->Stats.Speed;
	});
}

void ATurnManager::CheckCombatEnd()
{
	bool bPlayersDefeated = AreAllCharactersDefeated(PlayerCharacters);
	bool bEnemiesDefeated = AreAllCharactersDefeated(EnemyCharacters);

	if (bPlayersDefeated || bEnemiesDefeated)
	{
		// Clear any pending timers when combat ends
		GetWorld()->GetTimerManager().ClearTimer(NextTurnTimerHandle);
		CombatState = ECombatState::CombatEnded;
		OnCombatEnded.Broadcast();
	}
}

bool ATurnManager::AreAllCharactersDefeated(const TArray<AMedievalCharacter*>& Characters) const
{
	for (AMedievalCharacter* Character : Characters)
	{
		if (Character && Character->IsAlive())
		{
			return false;
		}
	}
	return true;
}
